File: day4/py/main.py
from time import perf_counter_ns

def partOne(lines):
    matches = []
    for line in lines:
        splits = line.rstrip('\n').split(':')[1].split('|')
        wins = set()
        scratch = set()

        split = list(filter(lambda x: (x != ""), splits[0].split(' ')))
        for n in split:
            wins.add(n)

        split = list(filter(lambda x: (x != ""), splits[1].split(' ')))
        for n in split:
            scratch.add(n)

        count = len(wins & scratch)
        if count > 0:
            matches.append(2 ** (count - 1))

    return sum(matches)

def partTwo(lines):
    lucky = []

    for line in lines:
        splits = line.rstrip('\n').split(':')[1].split('|')
        wins = set()
        scratch = set()

        split = list(filter(lambda x: (x != ""), splits[0].split(' ')))
        for n in split:
            wins.add(n)

        split = list(filter(lambda x: (x != ""), splits[1].split(' ')))
        for n in split:
            scratch.add(n)

        lucky.append(len(wins & scratch))
    
    cards = [1] * len(lucky)
    for i in range(len(cards)):
        for j in range(i + 1, i + 1 + lucky[i]):
            cards[j] += cards[i]
    
    return sum(cards)


lines = open("../input.txt", "r").readlines()

start = perf_counter_ns()
print("part one = ", partOne(lines))
print("time[us] = ", (perf_counter_ns() - start) // 1000)

start = perf_counter_ns()
print("part two = ", partTwo(lines))
print("time[us] = ", (perf_counter_ns() - start) // 1000)

File: day4/js/main.js
function partOne (lines) {
    let sum = 0
    for (let line of lines) {
        const splits = line.split(':')[1].split('|');
        const wins = new Set();
        const scratch = new Set();

        let part = splits[0].split(' ').filter((num) => num != "");
        for (let n of part) {
            wins.add(n);
        }
        part = splits[1].split(' ').filter((num) => num != "");
        for (let n of part) {
            scratch.add(n);
        }

        let count = 0;
        for (let num of wins) {
            if (scratch.has(num)) { count += 1; }
        }
        if (count > 0) {
            sum += 2**(count - 1);
        }
    }
    return sum;
}

function partTwo (lines) {
    const lucky = [];
    for (let line of lines) {
        const splits = line.split(':')[1].split('|');
        const wins = new Set();
        const scratch = new Set();

        let part = splits[0].split(' ').filter((num) => num != "");
        for (let n of part) {
            wins.add(n);
        }
        part = splits[1].split(' ').filter((num) => num != "");
        for (let n of part) {
            scratch.add(n);
        }

        let count = 0;
        for (let num of wins) {
            if (scratch.has(num)) { count += 1; }
        }
        lucky.push(count)
    }

    const cards = Array(lucky.length).fill(1);
    for (let i = 0; i < cards.length; i += 1) {
        for (let j = i + 1; j < i + 1 + lucky[i]; j += 1) {
            cards[j] += cards[i]
        }
    }
    let sum = 0;
    cards.map((val) => sum += val);
    return sum;
}

const fs = require('fs');
fs.readFile('../input.txt', 'utf8', function (err, data) {
    if (err) {
        console.error(err);
        return;
    }
    
    const input = data.split('\n');

    const start1 = performance.now();
    const res1 = partOne(input);
    const end1 = performance.now();
    console.log("part one =", res1);
    console.log("time[ms] =", end1 - start1);

    const start2 = performance.now();
    const res2 = partTwo(input);
    const end2 = performance.now();
    console.log("part two =", res2);
    console.log("time[ms] =", end2 - start2);
});